File: linda-core/src/main/java/it/unibo/coordination/linda/core/Template.java
package it.unibo.coordination.linda.core;

import java.util.function.Predicate;

public interface Template extends Predicate<Tuple> {

    default boolean matches(Tuple tuple) {
        return matchWith(tuple).isMatching();
    }

    <T extends Tuple, TT extends Template, K, V> Match<T, TT, K, V> matchWith(T tuple);

    @Override
    default boolean test(Tuple tuple) {
        return matches(tuple);
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/Template.kt
package it.unibo.coordination.linda.core

import java.util.function.Predicate

interface Template<T : Tuple<T>> : Predicate<T> {

    @JvmDefault
    fun matches(tuple: T): Boolean {
        return matchWith(tuple).isMatching
    }

    fun matchWith(tuple: T): Match<T, out Template<T>, *, *>

    @JvmDefault
    override fun test(tuple: T): Boolean {
        return matches(tuple)
    }
}